File: learning/zig/01_variables_types.zig
const std = @import("std");

pub fn main() !void {
    // ==========================================
    // 変数とデータ型
    // ==========================================

    std.debug.print("=== Zigの変数とデータ型 ===\n", .{});

    // ------------------------------------------
    // 変数の宣言
    // ------------------------------------------

    // const: 定数（変更不可）
    const pi: f32 = 3.14159;
    std.debug.print("定数 pi: {d}\n", .{pi});

    // var: 変数（変更可能）
    var counter: i32 = 0;
    std.debug.print("初期値 counter: {d}\n", .{counter});
    counter += 10;
    std.debug.print("変更後 counter: {d}\n", .{counter});

    // 型推論を使った宣言
    const auto_int = 42; // i32 として推論
    const auto_float = 3.14; // f64 として推論
    const auto_string = "Hello"; // *const [5:0]u8 として推論

    std.debug.print("型推論: int={d}, float={d}, string={s}\n", .{ auto_int, auto_float, auto_string });

    // ------------------------------------------
    // 整数型
    // ------------------------------------------

    std.debug.print("\n=== 整数型 ===\n", .{});

    // 符号付き整数
    const i8_val: i8 = -128; // 8bit: -128 to 127
    const i16_val: i16 = -32768; // 16bit: -32768 to 32767
    const i32_val: i32 = -2147483648; // 32bit
    const i64_val: i64 = -9223372036854775808; // 64bit

    // 符号なし整数
    const u8_val: u8 = 255; // 8bit: 0 to 255
    const u16_val: u16 = 65535; // 16bit: 0 to 65535
    const u32_val: u32 = 4294967295; // 32bit
    const u64_val: u64 = 18446744073709551615; // 64bit

    std.debug.print("i8: {d}, u8: {d}\n", .{ i8_val, u8_val });
    std.debug.print("i16: {d}, u16: {d}\n", .{ i16_val, u16_val });
    std.debug.print("i32: {d}, u32: {d}\n", .{ i32_val, u32_val });
    std.debug.print("i64: {d}, u64: {d}\n", .{ i64_val, u64_val });

    // カスタムビット幅整数
    const i3_val: i3 = -4; // 3bit符号付き: -4 to 3
    const u3_val: u3 = 7; // 3bit符号なし: 0 to 7
    std.debug.print("カスタムビット幅 i3: {d}, u3: {d}\n", .{ i3_val, u3_val });

    // ------------------------------------------
    // 浮動小数点型
    // ------------------------------------------

    std.debug.print("\n=== 浮動小数点型 ===\n", .{});

    const f16_val: f16 = 1.5; // 16bit浮動小数点
    const f32_val: f32 = 3.14159; // 32bit浮動小数点
    const f64_val: f64 = 2.71828182845904523536; // 64bit浮動小数点
    const f128_val: f128 = 1.41421356237309504880168872420969807856967187537694; // 128bit浮動小数点

    std.debug.print("f16: {d}\n", .{f16_val});
    std.debug.print("f32: {d}\n", .{f32_val});
    std.debug.print("f64: {d}\n", .{f64_val});
    std.debug.print("f128: {d}\n", .{f128_val});

    // ------------------------------------------
    // 論理型
    // ------------------------------------------

    std.debug.print("\n=== 論理型 ===\n", .{});

    const is_true: bool = true;
    const is_false: bool = false;

    std.debug.print("true: {}, false: {}\n", .{ is_true, is_false });

    // 論理演算
    const and_result = is_true and is_false; // false
    const or_result = is_true or is_false; // true
    const not_result = !is_true; // false

    std.debug.print("and: {}, or: {}, not: {}\n", .{ and_result, or_result, not_result });

    // ------------------------------------------
    // 文字・文字列型
    // ------------------------------------------

    std.debug.print("\n=== 文字・文字列型 ===\n", .{});

    // 文字（UTF-8コードポイント）
    const char_a: u8 = 'A';
    const char_unicode: u21 = '🦎'; // Zigのマスコット

    std.debug.print("文字 A: {c} (コード: {d})\n", .{ char_a, char_a });
    std.debug.print("Unicode文字: {u} (コード: {d})\n", .{ char_unicode, char_unicode });

    // 文字列リテラル
    const greeting = "Hello, Zig!";
    const multiline =
        \\これは複数行の
        \\文字列です。
        \\各行の先頭の\\は削除されます。
    ;

    std.debug.print("挨拶: {s}\n", .{greeting});
    std.debug.print("複数行:\n{s}\n", .{multiline});

    // ------------------------------------------
    // Optional型（nullable）
    // ------------------------------------------

    std.debug.print("\n=== Optional型 ===\n", .{});

    // Optional型は ? を使って宣言
    const maybe_number: ?i32 = 42;
    const empty_number: ?i32 = null;

    // Optional型のチェック
    if (maybe_number) |value| {
        std.debug.print("値があります: {d}\n", .{value});
    } else {
        std.debug.print("値がありません\n", .{});
    }

    if (empty_number) |value| {
        std.debug.print("値があります: {d}\n", .{value});
    } else {
        std.debug.print("値がありません（null）\n", .{});
    }

    // ------------------------------------------
    // 配列とスライス（基本）
    // ------------------------------------------

    std.debug.print("\n=== 配列とスライス（基本） ===\n", .{});

    // 固定長配列
    const numbers = [_]i32{ 1, 2, 3, 4, 5 };
    std.debug.print("配列の長さ: {d}\n", .{numbers.len});
    std.debug.print("最初の要素: {d}\n", .{numbers[0]});

    // スライス
    const slice = numbers[1..4]; // インデックス1から3まで
    std.debug.print("スライス: ", .{});
    for (slice) |value| {
        std.debug.print("{d} ", .{value});
    }
    std.debug.print("\n", .{});

    // ------------------------------------------
    // 型キャスト
    // ------------------------------------------

    std.debug.print("\n=== 型キャスト ===\n", .{});

    const small_int: i8 = 100;
    const big_int: i32 = small_int; // 暗黙的な拡張キャスト

    const big_value: i32 = 127;
    const small_value: i8 = @intCast(big_value); // 明示的な縮小キャスト（範囲チェックあり）

    std.debug.print("i8 -> i32: {d} -> {d}\n", .{ small_int, big_int });
    std.debug.print("i32 -> i8: {d} -> {d}\n", .{ big_value, small_value });

    // floatとintの変換
    const float_val: f32 = 3.7;
    const int_from_float: i32 = @intFromFloat(float_val);
    const float_from_int: f32 = @floatFromInt(int_from_float);

    std.debug.print("float -> int -> float: {d} -> {d} -> {d}\n", .{ float_val, int_from_float, float_from_int });
}

File: learning/rust/src/bin/01_variables_types.rs
fn main() {
    // ==========================================
    // 変数とデータ型
    // ==========================================

    println!("=== Rustの変数とデータ型 ===");

    // ------------------------------------------
    // 変数の宣言
    // ------------------------------------------

    println!("\n=== 変数の宣言 ===");

    // let: 不変変数（デフォルト）
    let name = "Alice";
    println!("不変変数 name: {}", name);

    // let mut: 可変変数
    let mut counter = 0;
    println!("初期値 counter: {}", counter);
    counter += 10;
    println!("変更後 counter: {}", counter);

    // 型注釈
    let age: u32 = 25;
    let height: f64 = 165.5;
    println!("年齢: {}, 身長: {}", age, height);

    // 変数のシャドーイング
    let x = 5;
    let x = x + 1; // 新しい変数として再定義
    let x = x * 2;
    println!("シャドーイングされた x: {}", x);

    // 型を変更するシャドーイング
    let spaces = "   ";
    let spaces = spaces.len(); // 文字列から数値に型変更
    println!("スペースの数: {}", spaces);

    // ------------------------------------------
    // 整数型
    // ------------------------------------------

    println!("\n=== 整数型 ===");

    // 符号付き整数
    let i8_val: i8 = -128; // 8bit: -128 to 127
    let i16_val: i16 = -32768; // 16bit: -32768 to 32767
    let i32_val: i32 = -2147483648; // 32bit (デフォルト)
    let i64_val: i64 = -9223372036854775808; // 64bit
    let i128_val: i128 = -170141183460469231731687303715884105728; // 128bit
    let isize_val: isize = -100; // ポインタサイズ

    println!("符号付き整数:");
    println!("i8: {}, i16: {}, i32: {}", i8_val, i16_val, i32_val);
    println!("i64: {}, i128: {}, isize: {}", i64_val, i128_val, isize_val);

    // 符号なし整数
    let u8_val: u8 = 255; // 8bit: 0 to 255
    let u16_val: u16 = 65535; // 16bit: 0 to 65535
    let u32_val: u32 = 4294967295; // 32bit
    let u64_val: u64 = 18446744073709551615; // 64bit
    let u128_val: u128 = 340282366920938463463374607431768211455; // 128bit
    let usize_val: usize = 100; // ポインタサイズ

    println!("符号なし整数:");
    println!("u8: {}, u16: {}, u32: {}", u8_val, u16_val, u32_val);
    println!("u64: {}, u128: {}, usize: {}", u64_val, u128_val, usize_val);

    // 数値リテラルの記法
    let decimal = 98_222; // 10進数（アンダースコアで区切り可能）
    let hex = 0xff; // 16進数
    let octal = 0o77; // 8進数
    let binary = 0b1111_0000; // 2進数
    let byte = b'A'; // バイトリテラル（u8のみ）

    println!("数値リテラル:");
    println!(
        "decimal: {}, hex: {}, octal: {}, binary: {}, byte: {}",
        decimal, hex, octal, binary, byte
    );

    // ------------------------------------------
    // 浮動小数点型
    // ------------------------------------------

    println!("\n=== 浮動小数点型 ===");

    let f32_val: f32 = 3.14159; // 32bit浮動小数点
    let f64_val: f64 = 2.718281828459045; // 64bit浮動小数点（デフォルト）

    println!("f32: {}, f64: {}", f32_val, f64_val);

    // 科学記法
    let scientific1 = 1e6; // 1,000,000
    let scientific2 = 2.5e-3; // 0.0025

    println!("科学記法: 1e6 = {}, 2.5e-3 = {}", scientific1, scientific2);

    // ------------------------------------------
    // 論理型
    // ------------------------------------------

    println!("\n=== 論理型 ===");

    let is_rust_awesome: bool = true;
    let is_learning: bool = false;

    println!("Rustは素晴らしい: {}", is_rust_awesome);
    println!("学習中: {}", is_learning);

    // 論理演算
    let and_result = is_rust_awesome && is_learning;
    let or_result = is_rust_awesome || is_learning;
    let not_result = !is_rust_awesome;

    println!(
        "AND: {}, OR: {}, NOT: {}",
        and_result, or_result, not_result
    );

    // ------------------------------------------
    // 文字型
    // ------------------------------------------

    println!("\n=== 文字型 ===");

    let char_a: char = 'A';
    let char_unicode: char = '🦀'; // Rustのマスコット（カニ）
    let char_japanese: char = 'あ';
    let char_emoji: char = '😊';

    println!(
        "文字: {}, Unicode: {}, 日本語: {}, 絵文字: {}",
        char_a, char_unicode, char_japanese, char_emoji
    );

    // ------------------------------------------
    // 文字列型
    // ------------------------------------------

    println!("\n=== 文字列型 ===");

    // 文字列スライス（&str）
    let string_slice: &str = "Hello, Rust!";
    let literal = "これは文字列リテラルです";

    println!("文字列スライス: {}", string_slice);
    println!("リテラル: {}", literal);

    // String型（所有権を持つ文字列）
    let mut owned_string = String::from("Hello");
    owned_string.push_str(", World!");

    println!("所有権を持つ文字列: {}", owned_string);

    // rawストリング
    let raw_string = r"このraw文字列では\nや\tをエスケープしません";
    let raw_multiline = r#"
        複数行のraw文字列
        "引用符"も使えます
    "#;

    println!("Raw文字列: {}", raw_string);
    println!("Raw複数行: {}", raw_multiline);

    // ------------------------------------------
    // タプル型
    // ------------------------------------------

    println!("\n=== タプル型 ===");

    let person: (String, u32, f64) = (String::from("Bob"), 30, 175.5);
    println!("人物タプル: {:?}", person);

    // タプルの分解
    let (name, age, height) = person;
    println!("名前: {}, 年齢: {}, 身長: {}", name, age, height);

    // インデックスによるアクセス
    let coordinates = (10.5, 20.3);
    println!("座標: x = {}, y = {}", coordinates.0, coordinates.1);

    // 空のタプル（unit型）
    let unit: () = ();
    println!("Unit型: {:?}", unit);

    // ------------------------------------------
    // 配列型
    // ------------------------------------------

    println!("\n=== 配列型 ===");

    // 固定長配列
    let numbers: [i32; 5] = [1, 2, 3, 4, 5];
    let zeros = [0; 3]; // [0, 0, 0]

    println!("数値配列: {:?}", numbers);
    println!("ゼロ配列: {:?}", zeros);
    println!("配列の長さ: {}", numbers.len());
    println!("最初の要素: {}", numbers[0]);

    // 配列のスライス
    let slice = &numbers[1..4]; // インデックス1から3まで
    println!("スライス: {:?}", slice);

    // ------------------------------------------
    // 型変換
    // ------------------------------------------

    println!("\n=== 型変換 ===");

    // as キーワードによる型変換
    let integer = 42i32;
    let float_from_int = integer as f64;

    println!("整数から浮動小数点: {} -> {}", integer, float_from_int);

    // 文字列から数値への変換
    let string_number = "42";
    let parsed_number: i32 = string_number.parse().unwrap();

    println!("文字列から数値: \"{}\" -> {}", string_number, parsed_number);

    // 数値から文字列への変換
    let number = 123;
    let number_string = number.to_string();

    println!("数値から文字列: {} -> \"{}\"", number, number_string);

    // ------------------------------------------
    // 定数
    // ------------------------------------------

    println!("\n=== 定数 ===");

    const MAX_POINTS: u32 = 100_000;
    const PI: f64 = 3.14159265359;

    println!("最大ポイント: {}", MAX_POINTS);
    println!("円周率: {}", PI);

    // ------------------------------------------
    // Option型とResult型（基本）
    // ------------------------------------------

    println!("\n=== Option型とResult型（基本） ===");

    // Option型
    let some_number: Option<i32> = Some(5);
    let no_number: Option<i32> = None;

    match some_number {
        Some(value) => println!("値があります: {}", value),
        None => println!("値がありません"),
    }

    match no_number {
        Some(value) => println!("値があります: {}", value),
        None => println!("値がありません"),
    }

    // Result型
    let division_result: Result<i32, &str> = divide(10, 2);
    match division_result {
        Ok(value) => println!("除算結果: {}", value),
        Err(error) => println!("エラー: {}", error),
    }

    let error_result: Result<i32, &str> = divide(10, 0);
    match error_result {
        Ok(value) => println!("除算結果: {}", value),
        Err(error) => println!("エラー: {}", error),
    }
}

fn divide(a: i32, b: i32) -> Result<i32, &'static str> {
    if b == 0 {
        Err("ゼロで割ることはできません")
    } else {
        Ok(a / b)
    }
}
